respringless_tweaks.md (#15)
NightwindDev/Tweak-Tutorial: respringless_tweaks.md

# How do you create a respringless tweak?

> * It's actually really easy if you understand the basics of it. What I'll make is a simple tweak that adds a blur to the LockScreen, with the option to change the its alpha/intensity on the fly. Think about it, how could that happen without a respring? Simple, Preferences needs to communicate with SpringBoard to notify that the blur value has been changed, so that it can refresh the UI. The `NSNotificationCenter` Foundation API makes that really straightforward.

Reviewer: ```suggestion
*What we'll make is a simple tweak that adds a blur to the lock screen, with the option to change the its alpha/intensity on the fly. How can this happen without a respring? The Settings app will need to communicate with SpringBoard to notify that the blur value has been changed, so that it can refresh the UI. The `NSNotificationCenter` Foundation API makes that really straightforward.
```

> I won't explain what `NSNotificationCenter` does or how it works in too much detail because if you've got this far then you should already know the basics of it, if not then I'd suggest reading its documentation.

Reviewer: ```suggestion
You can explore the `NSNotificationCenter` API's documentation here: https://developer.apple.com/documentation/foundation/nsnotificationcenter.
```

## Tweak

Open your terminal and create a tweak project, your `Tweak.x` should look like this:

```objc

> #import "Tweah.h"

Reviewer: ```suggestion
#import "Tweak.h"
```

static float blurIntensity;

static void loadPrefs(void) {

>     NSUserDefaults *prefs = [[NSUserDefaults alloc] initWithSuiteName: kSuiteName];

Reviewer: ```suggestion
    NSUserDefaults *prefs = [[NSUserDefaults alloc] initWithSuiteName:@"com.example.respringlesstweakprefs"];
```
Let's try to make the code as portable as possible.

    blurIntensity = [prefs objectForKey:@"blurIntensity"] ? [prefs floatForKey:@"blurIntensity"] : 0.85;

}

> static _UIBackdropView *blurView;

Reviewer: This should probably be a `%property`, not a global static variable.

%hook CSCoverSheetViewController

%new

- (void)setupBlur {

    _UIBackdropViewSettings *settings = [_UIBackdropViewSettings settingsForStyle:2];

    blurView = [[_UIBackdropView alloc] initWithFrame:CGRectZero autosizesToFitSuperview:YES settings:settings];
    blurView.alpha = blurIntensity;
    [self.view insertSubview:blurView atIndex:0];   

}

%new

- (void)updateBlurIntensity {

    loadPrefs();
    blurView.alpha = blurIntensity;

}

- (void)viewDidLoad {

    %orig;
    [self setupBlur];

>     [NSDistributedNotificationCenter.defaultCenter addObserver:self selector:@selector(updateBlurIntensity) name:RespringlessTweakDidUpdateBlurIntensityNotification object: nil];

Reviewer: Might want to show an alternative way of doing this which does not require a new selector, rather using a block.

}

%end

%ctor {
    loadPrefs();
}
```

Then your `Tweak.h` should have:

```objc
@import UIKit;

> #import "Common.h"

Reviewer: Might want to think about removing this file and adding it to the main `Tweak.h` for a more concise example project.

@interface _UIBackdropViewSettings : NSObject
+ (id)settingsForStyle:(NSInteger)style;
@end

@interface _UIBackdropView : UIView
- (id)initWithFrame:(CGRect)frame autosizesToFitSuperview:(BOOL)autosizes settings:(_UIBackdropViewSettings *)settings;
@end

@interface CSCoverSheetViewController: UIViewController
- (void)setupBlur;
@end
```

Finally, `Common.h`

```objc

> static NSString *const kSuiteName = @"me.luki.respringlesstweakprefs";

Reviewer: ```suggestion

```
Remove the `kSuiteName` constant

> static NSNotificationName const RespringlessTweakDidUpdateBlurIntensityNotification = @"RespringlessTweakDidUpdateBlurIntensityNotification";

Reviewer: Using a more specific notification name would be nicer, for example `com.example.respringlesstweakprefs/DidUpdateBlurIntensityNotification`. Also I think removing the constant is best

@interface NSDistributedNotificationCenter: NSNotificationCenter
@end

> ```

Reviewer: Probably should move all this to `Tweak.h`

## Explanation

> * First we create the `loadPrefs(void)` function that will contain our prefs, adding the `float` key for the blur intensity.

Reviewer: ```suggestion
* First we create the `void loadPrefs(void)` function that will contain our prefs, adding the `CGFloat` key for the blur intensity.
```
Specify the return type and use `CGFloat` since we are dealing with graphics
